### ISSUE TITLE
Reconsider merging exports and $foreign in psc-bundle

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#1239]
It's not a huge problem, but doing this makes all foreign functions public regardless of whether they're exported or not.

**Comment by paf31** [comment]
I'm fine with that, if whatever we replace it with works. The only downside I can see is that we end up making foreign modules special, and if we want to start using ES6 modules one day, we'd want to treat them as regular modules.

**Comment by garyb** [comment]
Yeah, I realised that it's not the easiest thing to fix, just thought it was worth mentioning as a not entirely consistent behaviour.

**Comment by texastoland** [comment]
@garyb Does this affect the Rollup stuff at all?

**Comment by garyb** [comment]
Possibly yeah, if we re-export everything in `$foreign` it means unused FFI definitions will not be erased.

**Comment by natefaubion** [comment]
Should this be closed?

**Comment by garyb** [comment]
I think it'd still be better to only export the public members of a module, regardless of whether they're coming from the FFI or not.

**Comment by rhendric** [comment]
This is no longer about our bundler, but it's still relevant, right? In the new ES modules regime, we explicitly reexport everything from the FFI, and that's still maybe not a great idea?

**Comment by JordanMartinez** [comment]
Good point. If the FFI file is an ES module, then should be able to just reuse it.
```javascript
const foo = function () { 
  console.log("this is not exported"); 
};
export const bar = function () {
  console.log("The answer:");
  foo();
};
```
```purescript
module Main where

foreign import bar :: Effect Unit

main = bar
```
should produce something like the following, right?
```javascript
// foreign.js
// (copied over as-is)
```
```javascript
// index.js
import { bar } from "./foreign.js";

export const main = bar;
```

**Comment by rhendric** [comment]
I believe that's the claim, and I support it.

I also think that the ES modules transition would be an excellent time to do this and minimize weird ABI-like breakage. If someone's importing PureScript modules directly from JavaScript, they could be depending on foreign exports being reexported from the main module, but they're going to have to rewrite their JavaScript code already to use ES imports so we might as well also ask them to change where they import things from.

**Comment by rhendric** [comment]
Actually, let me revise that a little. Given that the `Main` module doesn't have an explicit list of names to export, I believe that `bar` should be reexported. (`foo` of course should not be.) If we had `module Main (main) where` instead, I believe what you wrote is what should be produced.

Looking at what the compiler is currently doing with the `Effect` module (which has an explicit list of exports) and the `Effect.Console` module (which doesn't), I think we are actually already doing this; so if my revised perspective above is the consensus, we can close this issue. If it isn't the consensus, I would like to be persuaded on it.

**Comment by JordanMartinez** [comment]
> Given that the Main module doesn't have an explicit list of names to export, I believe that bar should be reexported. (foo of course should not be.)

Ah, right. That's what I meant. If an export list isn't defined, then everything aside from modules are re-exported. Modules can only be re-exported via syntax like
```purs
module Main (module ReExport) where

import Prelude (show) as ReExport
```
